### PR TITLE
Create a standalone Medusa deployment

### LINKED ISSUE #483: K8SSAND-1354 ⁃ Deploy a Medusa grpc pod for operator direct interactions
In order to generate restore mappings in the operator directly, we'll need it to be able to communicate with the storage backend without having Cassandra pods up.
An additional Medusa pod then needs to be deployed for that need.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1354) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1354
┆priority: Medium

### PR DESCRIPTION
**Pull request by adejanovski**
Allows the operator to communicate with the Medusa storage backend without requiring a Cassandra pod to be up.

**What this PR does**:
Creates a standalone Medusa deployment with the associated service.
Allows the operator to communicate with the Medusa storage backend without requiring a Cassandra pod to be up. This will be used to compute the restore mapping in the operator but also probably to handle synchronization of custom resources with backups in remote storage.

**Which issue(s) this PR fixes**:
Fixes #483

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by burmanm** [comment]
This seems to deploy the medusa to have a standalone instance always. Why do we need it always? Wouldn't we need it only when we intend to communicate with it?

**Comment by burmanm** [comment]
Well, I guess for synchronization it might make sense.

**Comment by jsanda** [comment]
Yes, it would be needed to sync with remote storage. I suppose we could consider (as an optimization) not deploying it if there are Cassandra pods with Medusa enabled. Not sure I am convinced that the pros would outweigh the cons.

File: controllers/k8ssandra/medusa_reconciler.go
Comment (jsanda) [inline controllers/k8ssandra/medusa_reconciler.go:152]: This doesn't make sense, at least not here in the K8ssandraClusterReconciler. The motivation of having a stand alone Medusa deployment is that the remote storage can be queried when there is no K8ssandraCluster. 

The work for this should happen as part of #495 which I have started but not gotten too far yet.

Comment (adejanovski) [reply controllers/k8ssandra/medusa_reconciler.go:152]: That's not exactly what I had understood. The motivation was to be able to communicate with the Medusa storage backend even if no Cassandra pods were up yet, in order to perform a restore on the very first start of these pods.
A K8ssandraCluster object is still our starting point IMO. It's where we'll define the name of the backup we want to restore, as well as the medusa backend settings.

**Comment by Miles-Garnsey** [comment]
Made these two review comments offline, putting them here so they aren't lost.

1. Can you please rebase? It looks like this branch is 20 commits behind and there is a conflict in `reconcile.go` which makes me think a review right now may not be too useful as the rebase will lead to quite a few changes.
2. Because we're deploying a reasonably substantial new resource (a new pod is a relatively big change IMO), can we please get some e2e or kuttl tests?

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [inline pkg/medusa/reconcile.go:192]: This function returns a lot of values, can we please use named returns to make it clearer + more self documenting?

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [inline pkg/medusa/reconcile.go:196]: This logging feels excessive.

File: test/e2e/medusa_test.go
Comment (Miles-Garnsey) [inline test/e2e/medusa_test.go:139]: Doesn't `ClusterKey` contain `namespace`? Do we need to pass them separately?

File: controllers/k8ssandra/medusa_reconciler_test.go
Comment (Miles-Garnsey) [inline controllers/k8ssandra/medusa_reconciler_test.go:97]: It might be good to check that the pod becomes ready here as well as just being created.

File: test/e2e/medusa_test.go
Comment (Miles-Garnsey) [reply test/e2e/medusa_test.go:139]: Also, can we put this in `pkg/test`? `suite_test.go` is too long and doesn't have much coherence.

File: controllers/k8ssandra/medusa_reconciler.go
Comment (Miles-Garnsey) [inline controllers/k8ssandra/medusa_reconciler.go:167]: actual and desired _deployments_?

File: controllers/k8ssandra/medusa_reconciler.go
Comment (Miles-Garnsey) [inline controllers/k8ssandra/medusa_reconciler.go:194]: This logic does not use the `CompareHashAnnotations` logic typically used in reconciliation. I'm open to not using it because `CompareHashAnnotations` doesn't actually allow for real self healing (the hash annotation is not updated when a user manually modifies an object, so an object can drift without being detected).

Just wondering why we're using two different methodologies really, I don't think it is necessarily a problem.

**Review by Miles-Garnsey — CHANGES_REQUESTED** [review]
High level feedback here:

* I think some refactoring is needed. It looks like there is similar logic for (at least) volume reconciliation operating on both the standalone deployment and the pod within the CassDC's STS. It would be good to see this logic unified in a single function. 
*  `medusaMainContainer` is another place I'd also like to see some more clarity. There are a few functions (including `medusaMainContainer`) where either the returns or the parameters might benefit from some attention.
* Packaging - it might be good to split functions in `pkg/medusa/reconcile` into multiple files. `reconcile.go` isn't very descriptive.
* Tests: I think unit tests might be desirable for fns like `medusaMainContainer`, `StandaloneMedusaDeployment`, `StandaloneMedusaService`.`UpdateMedusaVolumes`. E2E tests should probably check that the pod becomes ready rather than just being created.

I think once the function signatures are straightened out and you've considered a refactoring to unify the logic for configuring the standalone and in-DC Medusa `Containers`, we'll be in good shape.

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [inline pkg/medusa/reconcile.go:208]: I might be wrong here since it isn't quite clear where medusaContainer comes from (in the worst case it seems to be an empty newly allocated Container from medusaMainContainer). So apologies if this is a silly question...

Do we have liveness and readiness checks for this container?

File: controllers/k8ssandra/medusa_reconciler.go
Comment (jsanda) [reply controllers/k8ssandra/medusa_reconciler.go:69]: I haven't looked at the PR yet, but we should only need a volume for the Medusa configmap.

File: controllers/k8ssandra/medusa_reconciler.go
Comment (adejanovski) [reply controllers/k8ssandra/medusa_reconciler.go:211]: In the end I decided not to use the hash for the service. There's really nothing to configure there, so once it's created then we're fine. I'm reluctant to add more code to handle this, as it's not really necessary.

File: controllers/k8ssandra/medusa_reconciler.go
Comment (adejanovski) [reply controllers/k8ssandra/medusa_reconciler.go]: fixed

File: pkg/medusa/reconcile.go
Comment (adejanovski) [reply pkg/medusa/reconcile.go:178]: Let me know how it all looks now.

File: pkg/medusa/reconcile.go
Comment (adejanovski) [reply pkg/medusa/reconcile.go]: it does :) I removed it

File: pkg/medusa/reconcile.go
Comment (adejanovski) [reply pkg/medusa/reconcile.go]: That whole logic got refactored. Let me know how it looks.

File: controllers/k8ssandra/medusa_reconciler.go
Comment (Miles-Garnsey) [reply controllers/k8ssandra/medusa_reconciler.go:211]: Did we determine that changes to labels etc. would not cause an update of the service? As long as changes flowing from the CRs do not cause uncaught updates to the service I'm fine with not using a hash comparison.

**Review by Miles-Garnsey — CHANGES_REQUESTED** [review]
You're going to love me I'm afraid here @adejanovski, I have a few more comments... 

Firstly, this is looking much better and 90% of the changes have fixed the majority of the areas I'd queried. 

I think we still need to attend to the following areas:

1. Liveness/readiness, I've suggested three options there, but I think we should define a default (raise an issue if you want this in a separate PR, I'm open to that.)
2. We need to have resource requests/limits on the Medusa container, you can make this configurable like your readiness/liveness checks if you want.
3. `UpdateMedusaVolumes` is still a bit odd to me, can we try to make the behaviour consistent between additionalvolumes/volumes?
4. One question RE hashannotation comparisons. 

Otherwise, thanks for your efforts, we're getting very close.

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [reply pkg/medusa/reconcile.go:178]: Very much better! Resolving.

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [reply pkg/medusa/reconcile.go]: Resolving subsequent to refactoring.

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [reply pkg/medusa/reconcile.go]: Understood that this is required so we can use the same Medusa config as used in the Cass STS pods. Suggest further review at a later time but this is fine for a temporary solution.

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [reply pkg/medusa/reconcile.go:278]: Understood that this is required so we can use the same Medusa config as used in the Cass STS pods. Suggest further review at a later time but this is fine for a temporary solution.

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [reply pkg/medusa/reconcile.go]: Much better!

File: apis/medusa/v1alpha1/medusa_types.go
Comment (adejanovski) [reply apis/medusa/v1alpha1/medusa_types.go:149]: There are default probes defined [here](https://github.com/adejanovski/k8ssandra-operator/blob/add-medusa-controller-pod/pkg/medusa/reconcile.go#L206-L242), and merged with whatever the user sets in config.
I guess I should only give the ability to override the timeout and threshold settings. I'll come up with something that matches what we do elsewhere in the code.
Good point about the resource limits, I'll add the ability to set them and set a default.

File: controllers/k8ssandra/medusa_reconciler.go
Comment (adejanovski) [reply controllers/k8ssandra/medusa_reconciler.go:64]: > Can we get this all consistent by doing everything (both volumes + additionalvolumes) in-place to start with. That'll be a good first step.

As much as I would want this, it's not possible without going against your other requirement of working on the pod template spec. The additional volumes aren't defined in the `podTemplateSpec`, they're under `dcConfig.StorageProperties`.

I'll rename `UpdateMedusaVolumes()` to `GenerateMedusaVolumes()` and have it return both volumes and additional volumes, then add them to the pods in an external loop. At least that behavior will be consistent for both types of volumes.

File: pkg/medusa/reconcile.go
Comment (adejanovski) [reply pkg/medusa/reconcile.go:278]: 👍

File: controllers/k8ssandra/medusa_reconciler.go
Comment (adejanovski) [reply controllers/k8ssandra/medusa_reconciler.go:211]: > As long as changes flowing from the CRs do not cause uncaught updates to the service I'm fine with not using a hash comparison.

The K8ssandraCluster name is all that could trigger a breaking change here and it's not possible, so we're safe 👍

**Comment by adejanovski** [comment]
@Miles-Garnsey, tests are all passing now, let me know if we're good to go.

File: test/e2e/medusa_test.go
Comment (Miles-Garnsey) [reply test/e2e/medusa_test.go]: This isn't even in `suite_test.go`. Ignore me.

File: pkg/medusa/reconcile.go
Comment (Miles-Garnsey) [inline pkg/medusa/reconcile.go:240]: Could/should this be done via Goalesce? Note that I can't see `TerminationGracePeriodSeconds` in here, which may be an important configurable in some cases.

File: apis/medusa/v1alpha1/medusa_types.go
Comment (Miles-Garnsey) [reply apis/medusa/v1alpha1/medusa_types.go]: Carrying forward comments into new review now that some of this has been addressed.

File: controllers/k8ssandra/medusa_reconciler.go
Comment (Miles-Garnsey) [reply controllers/k8ssandra/medusa_reconciler.go]: Agreed. I don't like using both volumes and additionalvolumes but this will do for now, pending my investigation of what additionalVolumes is doing under the hood.

File: apis/medusa/v1alpha1/medusa_types.go
Comment (Miles-Garnsey) [inline apis/medusa/v1alpha1/medusa_types.go:155]: Can we drop `Settings` off the end of these names. This is a total nit, but I think that e.g. `ReadinessProbe` is sufficient.

**Review by Miles-Garnsey — COMMENTED** [review]
I've made a few changes but we are truly down to brass tacks here and I am negotiable on all of these. 

I do feel that the tests could be better separated between integration and unit tests, and I'd like to think carefully about defining `ProbeSettings` given the similarities to `corev1. Probe`.

The other comments are for your consideration only.

File: pkg/medusa/reconcile.go
Comment (adejanovski) [reply pkg/medusa/reconcile.go:24]: You're right, I need to cut out a Medusa release first.

File: apis/medusa/v1alpha1/medusa_types.go
Comment (adejanovski) [reply apis/medusa/v1alpha1/medusa_types.go:155]: Since we're not defining the whole probe (I'm leaving out the Exec part on purpose because it shouldn't be overridden), I think it's fair to leave "settings" at the end.

File: controllers/k8ssandra/medusa_reconciler_test.go
Comment (adejanovski) [reply controllers/k8ssandra/medusa_reconciler_test.go:134]: Honestly, as long as it's covered I think I'm happy with the way it is. I spent too much time already on this PR 😓

**Comment by Miles-Garnsey** [comment]
@adejanovski, as discussed, it does appear that the sanitized clusterName field is the point of reference for downstream resource names, so that's all good. Apologies for my error there.

I've raised a ticket RE using goalesce to disentangle things a bit here, but as it is an implementation detail I won't block merge on this point.

I tested once again, changing the k8s resource name, cluster name, and rack names (but leaving the DC name unchanged). 

That tests shows that I can use cqlsh to extract data added to the original cluster from the restored cluster. So I think this is good to merge. Approving once again. Nice work on getting this monster across the line!

**Review by Miles-Garnsey — APPROVED** [review]

**Comment by adejanovski** [comment]
Closing in favor of #980

**Pull request closed without merging**
